protocol/l1-upgrades.md

```diff
@@ -0,0 +1,305 @@
+
+# Problem Statement + Context
+
+Our upgrades have the following issues:
+
+1. They are divorced from the implementation, meaning the work of writing the upgrade script occurs
+   in `superchain-ops` apart from the work of writing the updated code in the `optimism` monorepo.
+2. The act of writing the upgrade script is onerous, requiring:
+   1. A bespoke script with custom assertions.
+   2. A bespoke Validations.md document
+3. The work required to perform the upgrade is difficult for signers, as they must work through the
+   Validations.md document. This work has taken up to an hour in the past.
+4. The effort described above will scale linearly with the number of OP Chains.
+
+# Proposed Solution
+
+Each new release will have its own OP Contracts Manager which can perform the following distinct
+actions:
+
+1. Deploy new proxies for a new OP Chain (this work is already covered by [link to Blaine's doc]())
+1. Upgrade all existing OP Chains.
+1. Deploy any newly added superchain-shared contracts.
+1. Upgrade existing superchain-shared contracts.
+
+Note that the above functionality also makes for a good set of milestones, each of which can and
+should be implemented incrementally.
+
+## Declarative storage management
+
+Taking inspiration from Chugsplash, we wish to separate the concern of what bytecode to run, and
+what to write to storage. However we also wish to avoid using the current `StorageSetter` pattern,
+which simply provides logic to "write anything anywhere" in storage.
+
+In order to achieve this, each contract will be refactored to inherit from a `__Layout` contract. By
+way of example, we take an initializable version of `SimpleStorage`, and refactor it into the set of
+contracts which would be required to specify a hypothetical Magma release, which adds a new variable
+to the contract.
+
+```solidity
+contract SimpleStorage is Initializable {
+  uint public foo;
+  bytes32 public bar; // new variable added for the Magma release.
+
+
+  /// @notice initialize is called whether for a fresh system or when upgrading an existing system.
+  function initialize(uint _foo, bytes32 bar) public initializer {
+    // chains being upgraded need to read the current value of foo, then write it again.
+    foo = _foo;
+    bar = _bar;
+  }
+}
+```
+
+This `SimpleStorage` will become:
+
+```solidity
+abstract contract SimpleStorageLayout {
+  uint public foo;
+  bytes32 public bar;
+}
+
+contract SimpleStoragePopulator is SimpleStorageLayout {
+  function initialize() public {
+    foo = 1;
+    bar = 2_000
+  }
+
+  function upgrade() public {
+    bar = 2_000;
+  }
+}
+
+contract SimpleStorage is SimpleStorageLayout {
+  // Whatever runtime logic is needed
+}
+```
+
+With this architecture, the OPCM contract (which will be a non-proxied singleton per release) would
+then either:
+
+1. Deploy a new contract by:
+   1. Deploying a new proxy
+   1. setting its implementation to the `SimpleStoragePopulator`
```

Comment:
What if a given upgrade requires no changes to the initialization logic and therefore no populator is needed, do we still do the step as a no-op or does it get skipped? Let's cover that case in this doc also

Reply:
We won't have a Populator, so in this case we just `upgrade` instead of `upgradeAndCall`